File: Csharp/LcdNumbers/DigitScaler.cs
using System.Collections.Generic;
using Sharpen;

namespace Org.Codecop.Lcdnumbers
{
	/// <summary>Scales x and y of digit lines.</summary>
	public class DigitScaler
	{
		private readonly ScalingRepeater repeater;

		public DigitScaler(ScalingRepeater repeater)
		{
			Objects.RequireNonNull(repeater);
			this.repeater = repeater;
		}

		public virtual IList<Digit> Scale(IList<Digit> digits, Scaling scaling)
		{
			Objects.RequireNonNull(digits);
			Objects.RequireNonNull(scaling);
			// return digits.map(d -> scale(d, scaling));
			IList<Digit> scaled = new AList<Digit>();
			foreach (Digit d in digits)
			{
				scaled.Add(Scale(d, scaling));
			}
			return scaled;
		}

		private Digit Scale(Digit digit, Scaling scaling)
		{
			Objects.RequireNonNull(digit);
			Objects.RequireNonNull(scaling);
			if (scaling.None())
			{
				return digit;
			}
			IList<Line> scaled = new AList<Line>();
			bool oddLine = true;
			foreach (Line line in digit.Lines())
			{
				Line scaledLine = line.ScaleHorizontal(repeater, scaling);
				if (oddLine)
				{
					scaled.Add(scaledLine);
				}
				else
				{
					Sharpen.Collections.AddAll(scaled, repeater.Repeat(scaledLine, scaling));
				}
				oddLine = !oddLine;
			}
			return digit.Scale(scaled);
		}
	}
}

File: PHP/src/LcdNumbers/DigitScaler.php
<?php
namespace LcdNumbers;

/**
 * Scales x and y of digit lines.
 */
class DigitScaler
{

    private $repeater;

    public function __construct(ScalingRepeater $repeater)
    {
        $this->repeater = $repeater;
    }

    public function scale(array $digits, Scaling $scaling)
    {
        $scaledDigits = [];
        foreach ($digits as $digit) {
            array_push($scaledDigits, $this->scaleDigit($digit, $scaling));
        }
        return $scaledDigits;
    }

    private function scaleDigit(Digit $digit, Scaling $scaling)
    {
        if ($scaling->none()) {
            return $digit;
        }

        $scaled = [];

        $oddLine = true;
        foreach ($digit->lines() as $line) {
            $scaledLine = $line->scaleHorizontal($this->repeater, $scaling);
            if ($oddLine) {
                array_push($scaled, $scaledLine);
            } else {
                array_merge($scaled, $this->repeater->repeat($scaledLine, $scaling));
            }
            $oddLine = !$oddLine;
        }

        return $digit->scale($scaled);
    }
}
